New function modbus_reply_callback()

frodete opened the pull request:
The modbus_reply_callback() function is modelled closely on modbus_reply(),
but should be used when more complex actions are required for handling
Modbus functions that what is possible with the current mapping scheme.

jbysewski committed:
Add bswap macro to compile on OSX

yegorich committed:
Change msinttypes URL to GitHub repository

msinttypes moved to GitHub, so change URL accordingly.

frodete committed:
New function modbus_reply_callback()

The modbus_reply_callback() function is modelled iclosely on modbus_reply(),
but should be used when more complex actions are required for handling
Modbus functions that what is possible with the current mapping scheme.

yegorich committed:
Add an option to disable tests compilation

Signed-off-by: Yegor Yefremov <yegorslists@googlemail.com>

stephane committed:
Add Jakob to AUTHORS with Contributor License Agreement

Commit:
Fix "wildcard address" in TCP IPv6

stephane committed:
Fix address range in random-test-client

Signed-off-by: Stéphane Raimbault <stephane.raimbault@gmail.com>

stephane committed:
Change API modbus_mapping_(offsets_new) to (new_start_address)

Related to 52ab1bbea760ed8eaca184f7d875a2f52a116d0f.
The arguments have been changed (see documentation).

https://groups.google.com/d/msg/libmodbus/aXO8nBzW4Ew/uVGTDmvvBAAJ

stephane committed:
Use new modbus_mapping_new_start_address in unit tests

stephane committed:
Add more tests for new modbus_mapping_new_start_address

stephane committed:
Improve documentation of modbus_report_slave_id

Inspired by
https://github.com/remakeelectric/libmodbus/commit/56b2526f5140a7e7efd0a59d5f2e5fb200f5a632

stephane committed:
Minor

stephane committed:
Update NEWS file

stephane committed:
Slight change to modbus_report_slave_id doc.

Thanks to Karl Palsson.

stephane committed:
Add links to new modbus_*_float_* functions in index

stephane committed:
Move setting of option inside the relevant conditional group

stephane committed:
Add ./configure.scan to .gitignore

stephane committed:
Add unit-tests.sh to tarball

stephane committed:
Fix small leak (64 bytes in TCP) in unit-test-client

stephane committed:
Github's contributing and issue template files

stephane committed:
Rewrite and rename README as README.md in tests/

stephane commented:
Thank you for your patch and the very good documentation.
The patch duplicates too many lines of code.

I think we can achieve the same behaviour with https://github.com/stephane/libmodbus/pull/323.

It would be nice if you could participate to this new feature (tests, docs, whatever), I'm intend to talk about that on the mailing list... stay tuned.

stephane committed:
Fix CID 69140 - Bad bit shift operation (coverity) in tests

stephane committed:
CID 69142 - Unchecked return value in unit-test-server

stephane committed:
CID 69145 - Argument cannot be negative in unit-test-server

stephane committed:
Fix typo in 3053bd0adb

frodete commented:
It has been a while since I coded this together for a project and only recently managed to push it.  I agree that there are some duplication in the code.  However, when I started to make this I did try to look at some way to reduce the duplication.  There were a few reasons why I ended up with what I did:

1) I didn't want to break existing code base.  Not a real reason - unless you'r on a deadline. :)
2) I didn't want to pollute the code with additional branches which could adversely affect the branch predictor (i.e. less code, but worse performance).
3) My initial attempts produced code which was much worse from a maintenance perspective.
4) All in all, the total number of identical code lines between the two functions are relatively low - there are probably more duplications internally in each functions, like the error handling.

I am no longer involved in the project which spawned this patch, but I will try to contribute in any way I can.

StalderT committed:
Update configure.ac

https://github.com/stephane/libmodbus/issues/248

StalderT committed:
Update modbus-data.c

Making all in src
  CC       modbus.lo
  CC       modbus-data.lo
modbus-data.c:51:5: attention : #warning "Fallback on C functions for bswap_16" [-Wcpp]
 #   warning "Fallback on C functions for bswap_16"
     ^
modbus-data.c:52:24: erreur: redefinition of ‘bswap_16’
 static inline uint16_t bswap_16(uint16_t x)

stephane committed:
DRY in modbus_reply by improving response_exception()

Thanks to Rüdiger Ranft for the idea.

stephane committed:
Add debug message on unknown function and new unit test

stephane committed:
Fix handling of invalid function code (closes #315)

Thanks to paperwork

stephane committed:
Fix wrong function name in debug message

stephane committed:
Rename raw_rep to raw_rsp in unit-test-client

stephane committed:
Another round of DRY in modbus_reply()

stephane committed:
Minor coding conventions on defines

stephane committed:
Improve ifdef around bswap_16 for __CYGWIN__

stephane committed:
C_PROG_RANLIB is rendered obsolete by LT_INIT

stephane committed:
Rewrite new unit test for invalid function code

frodete committed:
Update according to API changes in response_exception() in
commit 73a88a747c473a7427ef70f0351ff6f5370c3495.

frodete committed:
Add debug message on unknown function code.

frodete committed:
Add debug messages on illegal callback responses.

frodete committed:
Refactored code in modbus_reply_callback() to avoid duplication of
code.

frodete commented:
I tried pushing three changes to this fork to address your concerne, but it looks like I ended up with a merge mess.  Does the changes make sense or should I start from scratch?
